cmd/rosetta-dispatcher-server/main.go

```diff
@@ -0,0 +1,297 @@
+		}
+	}
+
+	return sporksTable, nil
+}
+
+func (s SporkList) Len() int {
+	return len(s)
+}
+
+func (s SporkList) Less(i, j int) bool {
+	return s[i].First < s[j].First
+}
+
+func (s SporkList) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+func (s SporkList) ServerForHeight(height uint64) (*middleware.ProxyTarget, error) {
+	// Assume sorted and more than zero items
+
+	first := s[0].First
+	last := s[len(s)-1].Last
+
+	if height < first {
+		return nil, fmt.Errorf("height %d below supported range %d - %d", height, first, last)
+	}
+
+	if height > last {
+		return nil, fmt.Errorf("height %d above supported range %d - %d", height, first, last)
+	}
+
+	for _, spork := range s {
+		if height >= spork.First && height <= spork.Last {
+			return spork.ProxyTaget, nil
+		}
+	}
+
+	return nil, fmt.Errorf("spork for height %d not found", height)
+}
+
+func NewFlowHeightAwareBalancer(sporkList SporkList) *FlowHeightAwareBalancer {
+
+	//targets := make([]*middleware.ProxyTarget, sporkList.Len())
+	//
+	//for i, spork := range sporkList {
+	//
+	//}
+
+	return &FlowHeightAwareBalancer{
+		//targets:   nil,
+		sporkList: sporkList,
+	}
+}
+
+type FlowHeightAwareBalancer struct {
+	//targets []*middleware.ProxyTarget
+	//mutex   sync.RWMutex
+	sporkList SporkList
+}
+
+func (f *FlowHeightAwareBalancer) AddTarget(target *middleware.ProxyTarget) bool {
+	// boilerplate from labstack/echi
+	//for _, t := range f.targets {
+	//	if t.Name == target.Name {
+	//		return false
+	//	}
+	//}
+	//f.mutex.Lock()
+	//defer f.mutex.Unlock()
+	//f.targets = append(f.targets, target)
+	//return true
+	return false
+}
+
+func (f *FlowHeightAwareBalancer) RemoveTarget(name string) bool {
+	//f.mutex.Lock()
+	//defer f.mutex.Unlock()
+	//for i, t := range f.targets {
+	//	if t.Name == name {
+	//		f.targets = append(f.targets[:i], f.targets[i+1:]...)
+	//		return true
+	//	}
+	//}
+	//return false
+	return false
+}
+
+// All requests which  have BlockID have them in the same field
+type BlockAwareRequest struct {
+	BlockID identifier.Block `json:"block_identifier,omitempty"`
+}
+
+func (f *FlowHeightAwareBalancer) Next(e echo.Context) *middleware.ProxyTarget {
+	blockAware := BlockAwareRequest{}
+
+	var reqBody []byte
+	if e.Request().Body != nil { // Read
+		reqBody, _ = ioutil.ReadAll(e.Request().Body)
+	}
+	e.Request().Body = ioutil.NopCloser(bytes.NewBuffer(reqBody)) // Reset
+
+	err := json.Unmarshal(reqBody, &blockAware)
+	if err != nil {
+		e.Error(err)
+	}
+
+	// no block ID = just use latest spork
+	// TODO - or maybe random?
+	blockHeight := uint64(0)
+	if blockAware.BlockID.Index != nil {
+		blockHeight = *blockAware.BlockID.Index
+	}
+	if blockHeight == 0 {
+		return f.sporkList[len(f.sporkList)-1].ProxyTaget
+	}
+
+	for _, spork := range f.sporkList {
+		if blockHeight >= spork.First && blockHeight <= spork.Last {
+			return spork.ProxyTaget
+		}
+	}
```

Comment:
The `SporkList` type should probably provide a utility method to return the target given a height.